File: Data Structures & Algorithms/count-number-of-islands/submission-0.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        visited = set()
        directions = [[0,-1], [0,1], [-1,0], [1,0]]
        rows = len(grid)
        cols = len(grid[0])
        islands = 0
        
        def dfs(r, c):
            if (r < 0 or r >= rows or
                c < 0 or c >= cols or
                (r,c) in visited or
                grid[r][c] == "0"):
                return 0

            visited.add((r,c))
            for dr, dc in directions:
                dfs(r + dr, c + dc)
            return 1

        for r in range(rows):
            for c in range(cols):
                islands += dfs(r,c)
                
        return islands

File: Data Structures & Algorithms/islands-and-treasure/submission-0.cpp
class Solution {
public:
    vector<pair<int,int>> directions = {{0,1}, {0,-1}, {1,0}, {-1,0}};
    void dfs(vector<vector<int>>& grid, int i, int j, int distance) {
        int ROWS = grid.size();
        int COLS = grid[0].size();
        if (i<0 || i>=ROWS || j<0 || j>=COLS || grid[i][j] < 0) return;
        if (distance > 0 && grid[i][j] <= distance) return;
        grid[i][j] = distance;
        for (const auto& [row, col] : directions) {
            int ni = i + row;
            int nj = j + col;
            dfs(grid, ni, nj, distance+1);
        }
        
    }


    void islandsAndTreasure(vector<vector<int>>& grid) {
        int ROWS = grid.size();
        int COLS = grid[0].size();
        for (int i=0; i<ROWS; i++) {
            for (int j=0; j<COLS; j++) {
                if (grid[i][j] == 0) dfs(grid,i,j,0);
            }
        }
    }
};
